Analysis/include/QwRootFile.h

```diff
@@ -455,9 +715,70 @@ class QwRootFile {
     void ls()     { if (fMapFile) fMapFile->ls();     if (fRootFile) fRootFile->ls(); }
     void Map()    { if (fRootFile) fRootFile->Map(); }
     void Close()  {
+
+      
+      // Debug: List trees before closing
+
+      for (auto iter = fTreeByName.begin(); iter != fTreeByName.end(); iter++) {
+
+        if (!iter->second.empty() && iter->second.front()) {
+          TTree* tree = iter->second.front()->GetTree();
+          if (tree) {
+            Long64_t entries = tree->GetEntries();
+
+          }
+        }
+      }
+      
+      // Check if we should make the file permanent - restore original logic
       if (!fMakePermanent) fMakePermanent = HasAnyFilled();
+      
+
+#ifdef HAS_RNTUPLE_SUPPORT      
+      // Close all RNTuples before closing the file
+      for (auto& pair : fNTupleByName) {
+        for (auto& ntuple : pair.second) {
+          if (ntuple) ntuple->Close();
+        }
+      }
+#endif // HAS_RNTUPLE_SUPPORT
+      
+      // CRITICAL FIX: Explicitly write all trees before closing!
+      if (fRootFile) {
+
+        for (auto iter = fTreeByName.begin(); iter != fTreeByName.end(); iter++) {
+          if (!iter->second.empty() && iter->second.front()) {
+            TTree* tree = iter->second.front()->GetTree();
+            if (tree && tree->GetEntries() > 0) {
+
+              tree->Write();
+            }
+          }
+        }
+      }
+      
+      // Close the file and handle renaming
+      if (fRootFile) {
+        TString rootfilename = fRootFile->GetName();
+        
+        fRootFile->Close();
+        // Note: Don't delete fRootFile here - let destructor handle it
+        
+        // Handle file renaming for temporary files
+        if (fUseTemporaryFile && fMakePermanent) {
+          int err = rename( rootfilename.Data(), fPermanentName.Data() );
+          if (err) {
+            QwWarning << "Couldn't rename " << rootfilename << " to " << fPermanentName << QwLog::endl;
+          } else {
+            QwMessage << "Was able to rename " << rootfilename << " to " << fPermanentName << QwLog::endl;
+            QwMessage << "Root file is " << fPermanentName << QwLog::endl;
+          }
+        }
+      }
```

Comment:
This should not be here, because it is done in the QwRootfile destructor.  Or if it needs to be here, we need to modify the destructor.